// neonctl/commands/auth.js
import { join } from 'node:path';
import { writeFileSync, existsSync, readFileSync } from 'node:fs';
import { TokenSet } from 'openid-client';
import { auth, refreshToken } from '../auth.js';
import { log } from '../log.js';
import { getApiClient } from '../api.js';
import { isCi } from '../env.js';
import { CREDENTIALS_FILE } from '../config.js';
export const command = 'auth';
export const aliases = ['login'];
export const describe = 'Authenticate';
export const builder = (yargs) => yargs.option('context-file', {
    hidden: true,
});
export const handler = async (args) => {
    await authFlow(args);
};
export const authFlow = async ({ configDir, oauthHost, clientId, apiHost, forceAuth, }) => {
    if (!forceAuth && isCi()) {
        throw new Error('Cannot run interactive auth in CI');
    }
    const tokenSet = await auth({
        oauthHost: oauthHost,
        clientId: clientId,
    });
    const credentialsPath = join(configDir, CREDENTIALS_FILE);
    await preserveCredentials(credentialsPath, tokenSet, getApiClient({
        apiKey: tokenSet.access_token || '',
        apiHost,
    }));
    log.info(`Saved credentials to ${credentialsPath}`);
    log.info('Auth complete');
    return tokenSet.access_token || '';
};
const preserveCredentials = async (path, credentials, apiClient) => {
    const { data: { id }, } = await apiClient.getCurrentUserInfo();
    const contents = JSON.stringify({
        ...credentials,
        user_id: id,
    });
    // correctly sets needed permissions for the credentials file
    writeFileSync(path, contents, {
        mode: 0o700,
    });
};
export const ensureAuth = async (props) => {
    if (props._.length === 0 || props.help) {
        return;
    }
    if (props.apiKey || props._[0] === 'auth') {
        props.apiClient = getApiClient({
            apiKey: props.apiKey,
            apiHost: props.apiHost,
        });
        return;
    }
    const credentialsPath = join(props.configDir, CREDENTIALS_FILE);
    if (existsSync(credentialsPath)) {
        try {
            const tokenSetContents = await JSON.parse(readFileSync(credentialsPath, 'utf8'));
            const tokenSet = new TokenSet(tokenSetContents);
            if (tokenSet.expired()) {
                log.debug('using refresh token to update access token');
                const refreshedTokenSet = await refreshToken({
                    oauthHost: props.oauthHost,
                    clientId: props.clientId,
                }, tokenSet).catch((err) => {
                    const typedErr = err && err instanceof Error ? err : undefined;
                    log.error('failed to refresh token\n%s', typedErr?.message);
                    process.exit(1);
                });
                props.apiKey = refreshedTokenSet.access_token || 'UNKNOWN';
                props.apiClient = getApiClient({
                    apiKey: props.apiKey,
                    apiHost: props.apiHost,
                });
                await preserveCredentials(credentialsPath, refreshedTokenSet, props.apiClient);
                return;
            }
            const token = tokenSet.access_token || 'UNKNOWN';
            props.apiKey = token;
            props.apiClient = getApiClient({
                apiKey: props.apiKey,
                apiHost: props.apiHost,
            });
            return;
        }
        catch (e) {
            if (e.code !== 'ENOENT') {
                // not a "file does not exist" error
                throw e;
            }
            props.apiKey = await authFlow(props);
        }
    }
    else {
        props.apiKey = await authFlow(props);
    }
    props.apiClient = getApiClient({
        apiKey: props.apiKey,
        apiHost: props.apiHost,
    });
};
