// neonctl/commands/projects.js
import { log } from '../log.js';
import { projectCreateRequest, projectUpdateRequest, } from '../parameters.gen.js';
import { writer } from '../writer.js';
import { psql } from '../utils/psql.js';
import { updateContextFile } from '../context.js';
import { getComputeUnits } from '../utils/compute_units.js';
export const PROJECT_FIELDS = [
    'id',
    'name',
    'region_id',
    'created_at',
];
const REGIONS = [
    'aws-us-west-2',
    'aws-ap-southeast-1',
    'aws-ap-southeast-2',
    'aws-eu-central-1',
    'aws-us-east-2',
    'aws-us-east-1',
    'azure-eastus2',
];
const PROJECTS_LIST_LIMIT = 100;
export const command = 'projects';
export const describe = 'Manage projects';
export const aliases = ['project'];
export const builder = (argv) => {
    return argv
        .usage('$0 projects <sub-command> [options]')
        .command('list', 'List projects', (yargs) => yargs.options({
        'org-id': {
            describe: 'List projects of a given organization',
            type: 'string',
        },
    }), async (args) => {
        await list(args);
    })
        .command('create', 'Create a project', (yargs) => yargs.options({
        name: {
            describe: projectCreateRequest['project.name'].description,
            type: 'string',
        },
        'region-id': {
            describe: `The region ID. Possible values: ${REGIONS.join(', ')}`,
            type: 'string',
        },
        'org-id': {
            describe: "The project's organization ID",
            type: 'string',
        },
        psql: {
            type: 'boolean',
            describe: 'Connect to a new project via psql',
            default: false,
        },
        database: {
            describe: projectCreateRequest['project.branch.database_name'].description,
            type: 'string',
        },
        role: {
            describe: projectCreateRequest['project.branch.role_name'].description,
            type: 'string',
        },
        'set-context': {
            type: 'boolean',
            describe: 'Set the current context to the new project',
            default: false,
        },
        cu: {
            describe: 'The number of Compute Units. Could be a fixed size (e.g. "2") or a range delimited by a dash (e.g. "0.5-3").',
            type: 'string',
        },
    }), async (args) => {
        await create(args);
    })
        .command('update <id>', 'Update a project', (yargs) => yargs.options({
        name: {
            describe: projectCreateRequest['project.name'].description,
            type: 'string',
        },
        'ip-allow': {
            describe: projectUpdateRequest['project.settings.allowed_ips.ips']
                .description,
            type: 'string',
            array: true,
            deprecated: "Deprecated. Use 'ip-allow' command",
        },
        'ip-primary-only': {
            describe: projectUpdateRequest['project.settings.allowed_ips.primary_branch_only'].description,
            type: 'boolean',
            deprecated: "Deprecated. Use 'ip-allow' command",
        },
        cu: {
            describe: 'The number of Compute Units. Could be a fixed size (e.g. "2") or a range delimited by a dash (e.g. "0.5-3").',
            type: 'string',
        },
    }), async (args) => {
        await update(args);
    })
        .command('delete <id>', 'Delete a project', (yargs) => yargs, async (args) => {
        await deleteProject(args);
    })
        .command('get <id>', 'Get a project', (yargs) => yargs, async (args) => {
        await get(args);
    });
};
export const handler = (args) => {
    return args;
};
const list = async (props) => {
    const getList = async (fn) => {
        const result = [];
        let cursor;
        let end = false;
        while (!end) {
            const { data } = await fn({
                limit: PROJECTS_LIST_LIMIT,
                org_id: props.orgId,
                cursor,
            });
            result.push(...data.projects);
            cursor = data.pagination?.cursor;
            log.debug('Got %d projects, with cursor: %s', data.projects.length, cursor);
            if (data.projects.length < PROJECTS_LIST_LIMIT) {
                end = true;
            }
        }
        return result;
    };
    const ownedProjects = getList(props.apiClient.listProjects);
    const sharedProjects = props.orgId
        ? undefined
        : getList(props.apiClient.listSharedProjects);
    const out = writer(props);
    out.write(await ownedProjects, {
        fields: PROJECT_FIELDS,
        title: 'Projects',
    });
    if (sharedProjects) {
        out.write(await sharedProjects, {
            fields: PROJECT_FIELDS,
            title: 'Shared with you',
        });
    }
    out.end();
};
const create = async (props) => {
    const project = {};
    if (props.name) {
        project.name = props.name;
    }
    if (props.regionId) {
        project.region_id = props.regionId;
    }
    if (props.orgId) {
        project.org_id = props.orgId;
    }
    project.branch = {};
    if (props.database) {
        project.branch.database_name = props.database;
    }
    if (props.role) {
        project.branch.role_name = props.role;
    }
    if (props.cu) {
        project.default_endpoint_settings = props.cu
            ? getComputeUnits(props.cu)
            : undefined;
    }
    const { data } = await props.apiClient.createProject({
        project,
    });
    if (props.setContext) {
        updateContextFile(props.contextFile, {
            projectId: data.project.id,
        });
    }
    const out = writer(props);
    out.write(data.project, { fields: PROJECT_FIELDS, title: 'Project' });
    out.write(data.connection_uris, {
        fields: ['connection_uri'],
        title: 'Connection URIs',
    });
    out.end();
    if (props.psql) {
        const connection_uri = data.connection_uris[0].connection_uri;
        const psqlArgs = props['--'];
        await psql(connection_uri, psqlArgs);
    }
};
const deleteProject = async (props) => {
    const { data } = await props.apiClient.deleteProject(props.id);
    writer(props).end(data.project, {
        fields: PROJECT_FIELDS,
    });
};
const update = async (props) => {
    const project = {};
    if (props.name) {
        project.name = props.name;
    }
    if (props.ipAllow || props.ipPrimaryOnly != undefined) {
        const { data } = await props.apiClient.getProject(props.id);
        const existingAllowedIps = data.project.settings?.allowed_ips;
        project.settings = {
            allowed_ips: {
                ips: props.ipAllow ?? existingAllowedIps?.ips ?? [],
                primary_branch_only: props.ipPrimaryOnly ??
                    existingAllowedIps?.primary_branch_only ??
                    false,
            },
        };
    }
    if (props.cu) {
        project.default_endpoint_settings = props.cu
            ? getComputeUnits(props.cu)
            : undefined;
    }
    const { data } = await props.apiClient.updateProject(props.id, {
        project,
    });
    writer(props).end(data.project, { fields: PROJECT_FIELDS });
};
const get = async (props) => {
    const { data } = await props.apiClient.getProject(props.id);
    writer(props).end(data.project, { fields: PROJECT_FIELDS });
};
